src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2412,135 +2462,107 @@ public override BoundNode VisitConditionalOperator(BoundConditionalOperator node
+                TypeWithState convertedAlternativeResult = default;
 
-                if (!isConstantFalse)
+                if (consequenceEndReachable)
                 {
                     convertedConsequenceResult = convertResult(
                         node.Consequence,
                         consequence,
                         consequenceConversion,
                         resultTypeWithAnnotations,
-                        consequenceResult);
+                        consequenceRValue);
                 }
 
-                if (!isConstantTrue)
+                if (alternativeEndReachable)
                 {
                     convertedAlternativeResult = convertResult(
                         node.Alternative,
                         alternative,
                         alternativeConversion,
                         resultTypeWithAnnotations,
-                        alternativeResult);
-                }
-
-                if (!convertedAlternativeResult.HasType)
-                {
-                    Debug.Assert(convertedConsequenceResult.HasType);
-                    visitResult = convertedConsequenceResult;
-                }
-                else if (!convertedConsequenceResult.HasType)
-                {
-                    Debug.Assert(convertedAlternativeResult.HasType);
-                    visitResult = convertedAlternativeResult;
-                }
-                else
-                {
-                    visitResult = TypeWithAnnotations.Create(resultType,
-                        convertedConsequenceResult.NullableAnnotation.Join(convertedAlternativeResult.NullableAnnotation));
+                        alternativeRValue);
                 }
-            }
-            else
-            {
-                NullableAnnotation resultNullableAnnotation;
 
-                if (isConstantTrue)
+                resultState = (consequenceEndReachable, alternativeEndReachable) switch
                 {
-                    resultNullableAnnotation = getNullableAnnotation(consequence, consequenceResult);
-                }
-                else if (isConstantFalse)
-                {
-                    resultNullableAnnotation = getNullableAnnotation(alternative, alternativeResult);
-                }
-                else
-                {
-                    resultNullableAnnotation = getNullableAnnotation(consequence, consequenceResult).Join(getNullableAnnotation(alternative, alternativeResult));
-                }
-
-                visitResult = TypeWithAnnotations.Create(node.Type.SetUnknownNullabilityForReferenceTypes(), resultNullableAnnotation);
+                    (false, false) => NullableFlowState.NotNull,
+                    (false, _) => convertedAlternativeResult.State,
+                    (_, false) => convertedConsequenceResult.State,
+                    _ => convertedConsequenceResult.State.Join(convertedAlternativeResult.State)
```

Comment:
>convertedConsequenceResult.State.Join(convertedAlternativeResult.State) [](start = 25, length = 71)

I think this whole switch expression can be replaced by `convertedConsequenceResult.State.Join(convertedAlternativeResult.State)` because the state of every expression in unreachable code in `NotNull`.

Reply:
Indeed, thanks. I made an adjustment to `VisitExpressionWithStackGuard` to that end (we should use reachability after the expression, rather than before, to adjust for result type when unreachable).